solidity/contracts/bridge/RedemptionWatchtower.sol

```diff
@@ -0,0 +1,618 @@
+            _watchtowerLifetime >= watchtowerLifetime,
+            "New lifetime must not be lesser than current one"
+        );
+
+        // Enforce the 5% hard cap.
+        require(
+            _vetoPenaltyFeeDivisor >= 20 || _vetoPenaltyFeeDivisor == 0,
+            "Redemption veto penalty fee must be in range [0%, 5%]"
+        );
+
+        // Enforce proper relationship between the delay levels. Use
+        // `>=` to allow for setting all delays to zero, if needed.
+        require(
+            _levelTwoDelay >= _levelOneDelay,
+            "Redemption level-two delay must not be lesser than level-one delay"
+        );
+        require(
+            _levelOneDelay >= _defaultDelay,
+            "Redemption level-one delay must not be lesser than default delay"
+        );
+
+        watchtowerLifetime = _watchtowerLifetime;
+        vetoPenaltyFeeDivisor = _vetoPenaltyFeeDivisor;
+        vetoFreezePeriod = _vetoFreezePeriod;
+        defaultDelay = _defaultDelay;
+        levelOneDelay = _levelOneDelay;
+        levelTwoDelay = _levelTwoDelay;
+        waivedAmountLimit = _waivedAmountLimit;
+
+        emit WatchtowerParametersUpdated(
+            _watchtowerLifetime,
+            _vetoPenaltyFeeDivisor,
+            _vetoFreezePeriod,
+            _defaultDelay,
+            _levelOneDelay,
+            _levelTwoDelay,
+            _waivedAmountLimit
+        );
+    }
+
+    /// @notice Determines whether a redemption request is considered safe.
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet that
+    ///        is meant to handle the redemption request.
+    /// @param redeemerOutputScript The redeemer's length-prefixed output
+    ///        script (P2PKH, P2WPKH, P2SH or P2WSH) that is meant to
+    ///        receive the redeemed amount.
+    /// @param balanceOwner The address of the Bank balance owner whose balance
+    ///        is getting redeemed.
+    /// @param redeemer The address that requested the redemption.
+    /// @return True if the redemption request is safe, false otherwise.
+    ///         The redemption is considered safe when:
+    ///         - The balance owner is not banned,
+    ///         - The redeemer is not banned,
+    ///         - There are no objections against past redemptions from the
+    ///           given wallet to the given redeemer output script.
+    function isSafeRedemption(
+        bytes20 walletPubKeyHash,
+        bytes calldata redeemerOutputScript,
+        address balanceOwner,
+        address redeemer
+    ) external view returns (bool) {
+        if (isBanned[balanceOwner]) {
+            return false;
+        }
+
+        if (isBanned[redeemer]) {
+            return false;
+        }
+
+        uint256 redemptionKey = Redemption.getRedemptionKey(
+            walletPubKeyHash,
+            redeemerOutputScript
+        );
+
+        if (vetoProposals[redemptionKey].objectionsCount > 0) {
+            return false;
+        }
```

Comment:
question regarding this block, what happens in case only one or two objections were raised and then delay is expired? 
I guess in that case redemption is still not safe but also new objections can't be raised?

Reply:
Such a redemption request is processed by the tBTC wallet after the delay. However, because there were objections to this redemption key (wallet + BTC address pair), the protocol blocks future redemptions from the same wallet to the given BTC address, just in case. Doing this is also easier from the redemption lifecycle management perspective and is consistent with what happens upon redemption timeout. That said:
- In case of the happy path (redemption processed without troubles) we allow reusing the given redemption key and ask the same wallet to redeem to the same BTC address
- In case of an unhappy path (objections raised, timeout hit) we block further usage of the given redemption key

I'm aware this may not be intuitive but this is a consequence of how tBTC identifies redemptions.

Reply:
Thank you for the explanation, I've found code where redemption handled after delay, could you point how redeemer banned in that case (one or two objections)?

Reply:
The redeemer is banned only after a full veto (3 objections). We don't ban after 1 or 2 objections.

Reply:
That I saw, whole question is about 1 or 2 objections, got it